recursion/reading-recursive-code.md

```diff
@@ -0,0 +1,288 @@
+
+The problem gets smaller by calling `pow` with a smaller exponent. As the `exponent` shrinks, it will eventually equal `0`!
+
+### Running `pow`
+
+How can we visualize the flow of code execution? Let's imagine the same `pow` function with several print statements, and then run it.
+
+```python
+def pow(num, exponent):
+    print("Calling pow! num:", num, "exponent:", exponent)
+    if exponent == 0:
+        return 1
+
+    return num * pow(num, exponent - 1)
+
+
+print(3, "raised to the power of", 5, "is", pow(3, 5))
+
+print("-------------")
+
+print(3, "raised to the power of", 6, "is", pow(3, 6))
+```
+
+We should get the following output:
+
+```
+Calling pow! num: 3 exponent: 5
+Calling pow! num: 3 exponent: 4
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 5 is 243
+-------------
+Calling pow! num: 3 exponent: 6
+Calling pow! num: 3 exponent: 5
+Calling pow! num: 3 exponent: 4
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 6 is 729
+```
+
+We can confirm that the `pow` function was called multiple times. Each time, the problem was made smaller, and `exponent` decremented by one. Eventually, when `exponent` was `0`, it met the conditions for the base case, and returned `1`.
+
+## Infinite Recursion
+
+A function that calls itself without working towards a base case will repeat infinitely.
+
+```python
+def infinite_recursion(n):
+    return n + infinite_recursion(n)
+```
+
+A function without a base case will never have a stopping point! Consider this code which has no base case:
+
+```python
+def infinite_recursion(n):
+    return n + infinite_recursion(n - 1)
+```
+
+Executing both of these functions results in a `RecursionError`!
+
+## Stacks and Function Calls
+
+A running program has a dedicated section of memory called **the system stack**.
+
+The system stack is where method calls and local variables are stored. It's responsible for keeping track of function calls.
+
+The "stack" in "system stack" implies that it uses the _stack_ data structure. The system stack keeps track of function calls using a stack data structure.
+
+### Intro to the Stack Data Structure
+
+How does a stack organize data? A stack organizes data like a stack of plates! When we add a plate to the stack of plates, we add it to the top. When we need a plate from the stack of plates, we remove it from the top.
+
+When a stack adds a new piece of data, it goes on top. When a stack removes one piece of data, it pops off the top item.
+
+### !callout-info
+
+## Stacks Are LIFO
+
+Stacks operate in a Last-In-First-Out (LIFO) manner. They remove things in the reverse order they were added.
+
+### !end-callout
+
```

Comment:
explain how stack frames are stored on the stack to prepare the reader for the mystery ahead
mention places we've already encountered the call stack
```suggestion

### Stack Frames Are Stored on the Stack

Each time we call a function, some tracking information is set up and pushed onto the stack. This collection of information is called a _stack frame_, or sometimes an _activation record_.

The exact structure and content of the stack frame varies from language to language, but a representative description of a stack frame would include:

- function arguments
- function local variables
- return value
- return location (where the function was called from)

![A box representing the collection of data stored with a stack frame. Contains four smaller boxes labelled arguments, locals, return value, and return location](../assets/recursion_reading-recursion_stack-frame.png)  
_Fig. Basic structure of a stack frame_

A stack frame is _pushed_ onto the call stack each time we call a function, and it is _popped_ off the call stack when the function completes.

![A pink region labelled Call Stack contains two stack frame boxes. An ellipsis is below the bottom stack frame to indicate there may be additional frames below these.](../assets/recursion_reading-recursion_frames.png)  
_Fig. The top of a call stack with two visible stack frames. Additional frames exist below these, indicated by the ellipsis._

The top of the call stack represents where the program is currently running. The frames below the current frame provide the sequence of nested function calls that led to the program being in the current function.

### !callout-info

## We Use the Call Stack When Debugging

There are two main ways we use the call stack while debugging.

1. We can view the call stack directly in the debugger. When we stop at a breakpoint, the debugging panel labelled `Call Stack` shows the sequence of function calls that led to the program arriving at that breakpoint.
2. When our programs crash, we see the call stack printed to the terminal. It shows the sequence of nested function calls that led to the program crashing.

### !end-callout

The call stack can become very deep and complicated. But with practice, we can get used to looking through the call stack to help understand what our programs are doing!

```